stan/math/rev/fun/rows_dot_product.hpp

```diff
@@ -17,12 +16,43 @@ template <typename Mat1, typename Mat2,
           require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
 inline Eigen::Matrix<var, Mat1::RowsAtCompileTime, 1> rows_dot_product(
     const Mat1& v1, const Mat2& v2) {
-  check_matching_sizes("dot_product", "v1", v1, "v2", v2);
-  Eigen::Matrix<var, Mat1::RowsAtCompileTime, 1> ret(v1.rows(), 1);
-  for (size_type j = 0; j < v1.rows(); ++j) {
-    ret.coeffRef(j) = dot_product(v1.row(j), v2.row(j));
+  check_matching_sizes("rows_dot_product", "v1", v1, "v2", v2);
+
+  ref_type_t<Mat1> v1_ref = v1;
+  ref_type_t<Mat2> v2_ref = v2;
+
+  arena_matrix<promote_scalar_t<double, Mat1>> arena_v1_val = value_of(v1_ref);
+  arena_matrix<promote_scalar_t<double, Mat2>> arena_v2_val = value_of(v2_ref);
+
+  arena_matrix<promote_scalar_t<var, Mat1>> arena_v1;
+  arena_matrix<promote_scalar_t<var, Mat2>> arena_v2;
+
+  if (!is_constant<Mat1>::value) {
+    arena_v1 = v1_ref;
+  }
+
+  if (!is_constant<Mat2>::value) {
+    arena_v2 = v2_ref;
   }
-  return ret;
+
+  Eigen::VectorXd out_val(arena_v1_val.rows());
+  for (size_t m = 0; m < arena_v1_val.rows(); ++m)
+    out_val.coeffRef(m) = arena_v1_val.row(m).dot(arena_v2_val.row(m));
```

Comment:
From [this](https://stackoverflow.com/a/53720369/2269255) stackoverflow it looks like the below is a hair faster

```suggestion
  Eigen::VectorXd out_val =  (arena_v1_val.cwiseProduct(arena_v2_val)).rowwise().sum();
```

Seems to have same result
https://godbolt.org/z/cqnvnv

Reply:
Odd, if I move that expression directly into the assignment like

```cpp
arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> out = (arena_v1_val.cwiseProduct(arena_v2_val)).rowwise().sum();
``
it throws an error from eigen

Reply:
I think `.sum()` is missing.